waiter/src/waiter/instance_tracker.clj

```diff
@@ -0,0 +1,99 @@
+
+  (handle-instances-event! [this instances]
+    (let [{:keys [clock handler-state]} this]
+      (log/info "newly failed instances" {:new-failed-instances instances})
+      (reset! handler-state {:last-error-time (clock)})))
+
+  (state [this _]
+    (let [{:keys [handler-state]} this]
+      @handler-state)))
+
+(defn create-instance-failure-event-handler
+  [{:keys [clock]}]
+  (let [handler-state (atom {:last-error-time nil})]
+    (DefaultInstanceFailureHandler. clock handler-state)))
+
+(defn start-instance-tracker
+  "Starts daemon thread that ingests various internal events and routes them to an external source"
+  [clock router-state-chan instance-failure-handler-component]
+  (cid/with-correlation-id
+    "instance-tracker"
+    (let [exit-chan (async/promise-chan)
+          query-chan (async/chan)
+          state-atom (atom {:all-failed-instances #{}
+                            :last-update-time nil})
+          ; create instance tracker component
+          query-state-fn
+          (fn instance-tracker-query-state-fn
+            [include-flags]
+            (let [{:keys [all-failed-instances last-update-time]} @state-atom]
+              (cond-> {:last-update-time last-update-time}
+                      (contains? include-flags "all-failed-instances")
+                      (assoc :all-failed-instances all-failed-instances)
+                      (contains? include-flags "instance-failure-handler")
+                      (assoc :instance-failure-handler (state instance-failure-handler-component [])))))
+          go-chan
+          (async/go
+            (try
+              (loop [{:keys [all-failed-instances] :as current-state} @state-atom]
+                (reset! state-atom current-state)
+                (let [[msg current-chan] (async/alts! [exit-chan router-state-chan query-chan] :priority true)
+                      next-state
+                      (condp = current-chan
+                        exit-chan
+                        (do
+                          (log/warn "stopping instance-tracker")
+                          (when (not= :exit msg)
+                            (throw (ex-info "Stopping instance-tracker" {:time (clock) :reason msg}))))
+
+                        router-state-chan
+                        (timers/start-stop-time!
+                          (metrics/waiter-timer "core" "instance-tracker" "router-state-chan")
+                          (let [{:keys [service-id->failed-instances]} msg
+                                all-failed-instances' (reduce (fn [cur-set cur-failed-instances]
+                                                                (set/union cur-set (set cur-failed-instances)))
+                                                              #{}
+                                                              (vals service-id->failed-instances))
+                                new-failed-instances (filter (complement all-failed-instances) all-failed-instances')]
```

Comment:
Should we compare failed instances by their ids in case other fields have changed?